client/network/src/lib.rs

```diff
@@ -454,3 +457,106 @@ async fn wait_to_announce<Block: BlockT>(
 		);
 	}
 }
+
+#[derive(Clone)]
+pub struct WaitForParachainTargetBlock<Block> {
+	phantom: PhantomData<Block>,
+}
+
+impl<Block: BlockT> WaitForParachainTargetBlock<Block> {
+	/// Get warp sync target block
+	pub async fn warp_sync_get(
+		para_id: ParaId,
+		relay_chain_interface: Arc<dyn RelayChainInterface>,
+		spawner: Arc<dyn SpawnNamed + Send + Sync>,
+	) -> Result<oneshot::Receiver<Block::Header>, BoxedError>
+	where
+		Block: BlockT + 'static,
+	{
+		let (sender, receiver) = oneshot::channel::<Block::Header>();
+
+		spawner.spawn(
+			"cumulus-parachain-wait-for-target-block",
+			None,
+			async move {
+				tracing::info!(
+					target: LOG_TARGET,
+					"waiting for target block in a background task...",
+				);
+				Self::wait_for_target_block(sender, para_id, relay_chain_interface).await;
+				tracing::info!(target: LOG_TARGET, "target block reached",);
+			}
+			.boxed(),
+		);
+
+		return Ok(receiver)
+	}
+
+	async fn wait_for_target_block(
+		sender: oneshot::Sender<Block::Header>,
+		para_id: ParaId,
+		relay_chain_interface: Arc<dyn RelayChainInterface>,
+	) {
+		let import_stream = relay_chain_interface.import_notification_stream();
+		match import_stream.await {
+			Ok(mut import_stream_notification) => match import_stream_notification.next().await {
+				Some(_header) => loop {
+					let is_syncing = relay_chain_interface
+						.is_major_syncing()
+						.await
+						.map_err(|e| {
+							tracing::error!(
+								target: LOG_TARGET,
+								"Unable to determine sync status. {}",
+								e
+							)
+						})
+						.unwrap_or(false);
+
+					if !is_syncing {
+						let mut finalized_heads =
+							match relay_chain_interface.finalized_heads(para_id).await {
+								Ok(finalized_heads_stream) => finalized_heads_stream,
+								Err(err) => {
+									tracing::error!(target: LOG_TARGET, error = ?err, "Unable to retrieve finalized heads stream.");
+									return
+								},
+							};
+
+						let finalized_head = if let Some(h) = finalized_heads.next().await {
+							h
+						} else {
+							tracing::debug!(
+								target: LOG_TARGET,
+								"Stopping following finalized head."
+							);
+							return
+						};
+
+						let target_header = match Block::Header::decode(&mut &finalized_head[..]) {
+							Ok(header) => header,
+							Err(err) => {
+								tracing::debug!(
+									target: LOG_TARGET,
+									error = ?err,
+									"Could not decode parachain header while following finalized heads.",
+								);
+								continue
+							},
+						};
+
+						let _ = sender.send(target_header);
+						break
+					}
+					tracing::info!(
+						target: LOG_TARGET,
+						"waiting for relay chain sync to complete......",
+					);
+					sleep(Duration::from_secs(120));
```

Comment:
```suggestion
```

You are sleeping in a future. This is wrong. I'm also not sure why you want to sleep here?

Reply:
There are async ways to "sleep".

Reply:
Now I see why you want to sleep, because of this loop. But this loop should be around this entire code.